packages/dds/tree/src/test/shared-tree/fuzzEditGenerator.ts

```diff
@@ -0,0 +1,331 @@
+        const insert: FuzzInsert = {
+            fuzzType: "insert",
+            parent: path,
+            field: nodeField,
+            index: nodeIndex,
+            value: state.random.integer(Number.MIN_SAFE_INTEGER, Number.MIN_SAFE_INTEGER),
+        };
+        return insert;
+    }
+
+    async function deleteGenerator(state: EditState): Promise<FuzzDelete> {
+        const trees = state.testTreeProvider.trees;
+        const tree = trees[state.treeIndex];
+        // generate edit for that specific tree
+        const path = containsAtLeastOneNode(tree)
+            ? getExistingRandomNodePosition(tree, state.random)
+            : undefined;
+        return {
+            fuzzType: "delete",
+            path,
+        };
+    }
+
+    async function setPayloadGenerator(state: EditState): Promise<FuzzSetPayload> {
+        const trees = state.testTreeProvider.trees;
+        const tree = trees[state.treeIndex];
+        // generate edit for that specific tree
+        const path = containsAtLeastOneNode(tree)
+            ? getExistingRandomNodePosition(tree, state.random)
+            : undefined;
+        return {
+            fuzzType: "setPayload",
+            path,
+            value: state.random.integer(Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER),
+        };
+    }
+
+    /**
+     * currently the acceptance conditions are based on some arbitrary number of edits.
+     * This should be changed to the max number of nodes allowed in tree.
+     * TODO: create helper function to get number of nodes in tree and use that to set acceptance condition
+     * */
+    const baseEditGenerator = createWeightedAsyncGenerator<FuzzChange, EditState>([
+        [insertGenerator, 5],
+        [deleteGenerator, 1],
+        [setPayloadGenerator, 1],
+    ]);
+
+    return async (state: FuzzTestState): Promise<Operation | typeof done> => {
+        const trees = state.testTreeProvider.trees;
+        // does not include last tree, as we want a passive client
+        const treeIndex = state.random.integer(0, trees.length - 2);
+        const contents = await baseEditGenerator({
+            ...state,
+            treeIndex,
+        });
+        state.numberOfEdits += 1;
+        if (contents === done) {
+            return done;
+        }
+        return { type: "edit", contents, index: treeIndex };
+    };
+};
+
+export function makeOpGenerator(): AsyncGenerator<Operation, FuzzTestState> {
+    const opWeights: AsyncWeights<Operation, FuzzTestState> = [
+        [makeEditGenerator(), 3],
+        [{ type: "synchronize" }, 1],
+    ];
+    return createWeightedAsyncGenerator(opWeights);
+}
+
+export interface NodeLocation {
+    parent: UpPath | undefined;
+    field: FieldKey;
+    index: number;
+}
+
+const moves = {
+    field: ["enterNode", "nextField"],
+    nodes: ["stop", "firstField"],
+};
+
+export function getRandomNodePosition(tree: ISharedTree, random: IRandom): NodeLocation {
```

Comment:
It's not very clear from `getRandomNodePosition` that its output may not correspond the path of an existing node. I recommend adopting the Place terminology: `getRandomPlace`.

Reply:
This function is quite complicated. It would be nice if this function could be implemented in terms of `getExistingRandomNodePosition`. I'm thinking something like this:
1. Check if the tree is empty. If it is, then return `{ parent: undefined, field: rootFieldKeySymbol, index: 0 }` like you're doing now.
2. Call `getExistingRandomNodePosition` to get a path to a random existing node.
3. Randomly choose from three options:
  (a) return the path to the node (i.e., return the place that describes the position before the node)
  (b) return the path to the node with with the index incremented by one (i.e., return the place that describes the position after the node)
  (c) return a new path that points to the start of a field below the node (something like `{ parent: path, field: testerKey, index: 0 }` where `path` is what you got from `getExistingRandomNodePosition`).

Reply:
Also it looks like it doesn't need to be exported.